derive/src/encode.rs

```diff
@@ -36,9 +34,22 @@ fn encode_fields<F>(
 {
 	let recurse = fields.iter().enumerate().map(|(i, f)| {
 		let field = field_name(i, &f.ident);
+		let encoded_as = utils::get_encoded_as_type(f);
+		let compact = utils::get_enable_compact(f);
 
-		quote_spanned! { f.span() =>
-			#dest.push(#field);
+		if encoded_as.is_some() && compact {
+			panic!("`encoded_as` and `compact` can not be used at the same time!");
+		}
+
+		if compact {
+			let field_type = &f.ty;
+			quote_spanned! { f.span() => { #dest.push(&Compact::<#field_type>::from(#field)); } }
```

Comment:
It would be a good idea to add comment on what particular `push` is called, since it's quite ambigious. Another option would be to rewrite the whole line as `<#dest as Output>::push(#field)` which would be self explanatory.